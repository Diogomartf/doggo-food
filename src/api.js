import axios from 'axios';

export const axiosApi = axios.create({
  headers: {
    'X-Requested-With': 'XMLHttpRequest',
    'Content-Type': 'application/json;charset=utf-8',
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Methods': 'GET'
  },
  baseURL: 'http://www.recipepuppy.com/api/'
});

const api = {
  recipes: {
    list: async ({ ingredients }) =>
      axiosApi.get('/', { params: { i: ingredients } })
  }
};

export default api;
