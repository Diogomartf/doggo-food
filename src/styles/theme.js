import colors from './colors';

export default {
  name: 'default theme',
  colors
};
