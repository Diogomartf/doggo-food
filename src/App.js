import React from 'react';
import Header from './components/Header';
import Recipes from './components/Recipes';
import api from './api';

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      searchText: '',
      recipes: [],
      recentSearches: [],
      inputFocus: false
    };
  }

  handleChange = search => {
    this.setState({ searchText: search });
  };

  handleSubmit = e => {
    e.preventDefault();
    const { searchText, recentSearches } = this.state;
    const hasMoreThan3Chars = searchText.length > 2;

    const updatedRecentSearches =
      recentSearches.length >= 5
        ? recentSearches.splice(1, 4).concat(searchText)
        : recentSearches.concat(searchText);

    if (hasMoreThan3Chars)
      api.recipes
        .list({ ingredients: searchText })
        .then(response =>
          this.setState({
            recipes: response.data.results,
            recentSearches: updatedRecentSearches
          })
        )
        .catch(response => console.log('error:', response));
  };

  onClickRecentSearch = search => {
    api.recipes
      .list({ ingredients: search })
      .then(response =>
        this.setState({
          recipes: response.data.results,
          searchText: search,
          inputFocus: false
        })
      )
      .catch(response => console.log('error:', response));
  };

  deleteRecentSearch = index => {
    const { recentSearches } = this.state;
    recentSearches.splice(index, 1);

    this.setState({ recentSearches: recentSearches });
  };

  render() {
    const { inputFocus, recentSearches } = this.state;

    return (
      <div>
        <Header
          inputFocus={inputFocus}
          recentSearches={recentSearches}
          handleSubmit={this.handleSubmit}
          handleChange={this.handleChange}
          searchText={this.state.searchText}
          handleOnFocus={this.handleOnFocus}
          handleOnBlur={this.handleOnBlur}
          deleteRecentSearch={this.deleteRecentSearch}
          onClickRecentSearch={this.onClickRecentSearch}
        />
        <Recipes
          recipes={this.state.recipes}
          searchText={this.state.searchText}
        />
      </div>
    );
  }
}

export default App;
