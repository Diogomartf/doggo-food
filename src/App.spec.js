import React from 'react';
import { shallow, mount } from 'enzyme';
import { ThemeProvider } from 'styled-components';
import theme from './styles';
import api from './api';

import App from './App';

const mockData = {
  data: {
    results: [
      {
        title: 'Salmon And Pesto Pasta Salad',
        href:
          'http://www.bigoven.com/130421-Salmon-And-Pesto-Pasta-Salad-recipe.html',
        ingredients:
          'salmon, pasta (in general), cherry tomato, pesto, olive oil, white wine vinegar, pesto',
        thumbnail: ''
      },
      {
        title: 'Fresh Pesto Pasta Salad',
        href:
          'http://find.myrecipes.com/recipes/recipefinder.dyn?action=displayRecipe&recipe_id=258269',
        ingredients:
          'pasta (in general), red wine vinegar, sugar, black pepper, salt, dijon mustard, garlic, olive oil, basil, parmesan cheese, pine nuts, grape tomatoes',
        thumbnail: 'http://img.recipepuppy.com/544602.jpg'
      },
      {
        title:
          'Whole-Wheat Pasta Shells with Spicy Tomato Pesto and Winter Greens',
        href:
          'http://find.myrecipes.com/recipes/recipefinder.dyn?action=displayRecipe&recipe_id=521883',
        ingredients:
          'pesto, water, sun dried tomato, almonds, parmesan cheese, basil, garlic, salt, red pepper, olive oil, salad greens, olive oil, onions, swiss chard, water, salt, black pepper, pasta, pasta (in general), parmesan cheese',
        thumbnail: 'http://img.recipepuppy.com/549528.jpg'
      }
    ]
  }
};

describe('<App/>', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  // it('should mount correctly', () => {
  //   wrapper = mount(
  //     <ThemeProvider theme={theme}>
  //       <App />
  //     </ThemeProvider>
  //   );

  //   expect(wrapper).toMatchSnapshot();
  // });

  it('handleChange should change state', () => {
    wrapper.instance().handleChange('pizza');
    expect(wrapper.state().searchText).toEqual('pizza');
  });

  it('should call handleSubmit correctly', () => {
    wrapper.setState({ searchText: 'pesto' });
    api.recipes.list = jest.fn().mockResolvedValue(mockData);

    const mockFormData = {
      preventDefault: jest.fn()
    };

    wrapper.instance().handleSubmit(mockFormData);

    expect(api.recipes.list).toHaveBeenCalledWith({ ingredients: 'pesto' });
  });

  it('should not call api when search text has less than 3 chars', () => {
    wrapper.setState({ searchText: 'pe' });
    api.recipes.list = jest.fn();

    const mockFormData = {
      preventDefault: jest.fn()
    };

    wrapper.instance().handleSubmit(mockFormData);

    expect(api.recipes.list).not.toHaveBeenCalled();
  });

  it('onClickRecentSearch should call api with recent search', () => {
    api.recipes.list = jest.fn().mockResolvedValue(mockData);
    wrapper.instance().onClickRecentSearch('pesto');

    expect(api.recipes.list).toHaveBeenCalledWith({ ingredients: 'pesto' });
  });

  it('deleteRecentSearch should remove select element from recentSearches', () => {
    wrapper.setState({ recentSearches: ['pesto', 'cheese', 'garlic'] });
    wrapper.instance().deleteRecentSearch(1);

    expect(wrapper.state().recentSearches).toEqual(['pesto', 'garlic']);
  });
});
