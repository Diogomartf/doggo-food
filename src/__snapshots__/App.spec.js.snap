// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> should render correctly 1`] = `
<div>
  <Header
    deleteRecentSearch={[Function]}
    handleChange={[Function]}
    handleSubmit={[Function]}
    onClickRecentSearch={[Function]}
    recentSearches={Array []}
    searchText=""
  />
  <Recipes
    recipes={Array []}
    searchText=""
  />
</div>
`;
